src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -623,74 +588,142 @@ public final boolean hasStaticInitializerForSerialization(Class<?> cl) {
     // Internals only below this point
     //
 
+    // Package-private to be accessible to NativeMethodAccessorImpl
+    // and NativeConstructorAccessorImpl
     static int inflationThreshold() {
-        return inflationThreshold;
+        return Config.instance().inflationThreshold;
     }
 
     static boolean noInflation() {
-        return noInflation;
+        return Config.instance().noInflation;
     }
 
     static boolean useMethodHandleAccessor() {
-        return (useDirectMethodHandle & METHOD_MH_ACCESSOR) == METHOD_MH_ACCESSOR;
+        return (Config.instance().useDirectMethodHandle & Config.METHOD_MH_ACCESSOR) == Config.METHOD_MH_ACCESSOR;
     }
 
     static boolean useFieldHandleAccessor() {
-        return (useDirectMethodHandle & FIELD_MH_ACCESSOR) == FIELD_MH_ACCESSOR;
+        return (Config.instance().useDirectMethodHandle & Config.FIELD_MH_ACCESSOR) == Config.FIELD_MH_ACCESSOR;
     }
 
     static boolean useNativeAccessorOnly() {
-        return useNativeAccessorOnly;
-    }
-
-    /** We have to defer full initialization of this class until after
-        the static initializer is run since java.lang.reflect.Method's
-        static initializer (more properly, that for
-        java.lang.reflect.AccessibleObject) causes this class's to be
-        run, before the system properties are set up. */
-    private static void checkInitted() {
-        if (initted) return;
-
-        // Defer initialization until module system is initialized so as
-        // to avoid inflation and spinning bytecode in unnamed modules
-        // during early startup.
-        if (!VM.isModuleSystemInited()) {
-            return;
-        }
+        return Config.instance().useNativeAccessorOnly;
+    }
 
-        Properties props = GetPropertyAction.privilegedGetProperties();
-        String val = props.getProperty("sun.reflect.noInflation");
-        if (val != null && val.equals("true")) {
-            noInflation = true;
-        }
+    private static boolean disableSerialConstructorChecks() {
+        return Config.instance().disableSerialConstructorChecks;
+    }
 
-        val = props.getProperty("sun.reflect.inflationThreshold");
-        if (val != null) {
-            try {
-                inflationThreshold = Integer.parseInt(val);
-            } catch (NumberFormatException e) {
-                throw new RuntimeException("Unable to parse property sun.reflect.inflationThreshold", e);
+    /**
+     * The configurations exist as an object to avoid race conditions.
+     * See bug 8261407. The object methods backed by indy may not be available.
+     */
```

Comment:
This comment is a bit unclear.   With the suggestion of moving the static `ReflectionFactory::config()` method and the comment I suggest below, I think that should be adequate to understand.  If not, we should improve the comment rather than relying on the readers to read the bug report.

Maybe the comment can be:
```
Configuration for core reflection which is configurable via system properties.
The user-configured settings are not available during early VM startup.

Note that the object methods of this Config record are called via indy which is 
available to use after initPhase1.   We can workaround that limitation by
implementing the object methods.
```

Reply:
> Note that the object methods of this Config record are called via indy which is 
> available to use after initPhase1.   We can workaround that limitation by
> implementing the object methods.

"Note that the object methods of this Config record are called via indy" is not very precise. They are called as any other method AFAIU. It's their implementation that uses indy, right? So perhaps, a better wording:

"Note that synthetic object methods of this Config record (toString, equals, hashCode) use indy in their implementation which is available to use after initPhase1.  These methods are currently not used, but should they be needed, a workaround is to explicitly (re)implement them."

Reply:
"uses indy" is the right description.   Minor tweak:  s/synthetic/the default implementation of/